src/web3/KeychainLock.ts

```diff
@@ -0,0 +1,254 @@
+          now
+        )
+        const createdAt = new Date(maxCreationTime)
+        accounts = [
+          { address: normalizedMnemonicAccountAddress, createdAt, importFromMnemonic: true },
+          ...accounts,
+        ]
+      }
+    }
+  } catch (e) {
+    Logger.error(`${TAG}@listStoredAccounts`, 'Error listing accounts', e)
+    throw new Error(ErrorMessages.KEYCHAIN_FETCH_ACCOUNTS)
+  }
+  return accounts
+}
+
+async function importAndStorePrivateKeyFromMnemonic(account: KeychainAccount, password: string) {
+  const mnemonic = await getStoredMnemonic(account.address, password)
+  if (!mnemonic) {
+    throw new Error('No mnemonic found in storage')
+  }
+
+  const { privateKey } = await generateKeysFromMnemonic(mnemonic)
+  if (!privateKey) {
+    throw new Error('Failed to generate private key from mnemonic')
+  }
+
+  // Prefix 0x here or else the signed transaction produces dramatically different signer!!!
+  const normalizedPrivateKey = normalizeAddressWith0x(privateKey)
+  const accountFromPrivateKey = normalizeAddressWith0x(privateKeyToAddress(normalizedPrivateKey))
+  if (accountFromPrivateKey !== account.address) {
+    throw new Error(
+      `Generated private key address (${accountFromPrivateKey}) does not match the existing account address (${account.address})`
+    )
+  }
+
+  await storePrivateKey(normalizedPrivateKey, account, password)
+
+  return normalizedPrivateKey
+}
+
+export async function clearStoredAccounts() {
+  Logger.info(`${TAG}@clearStoredAccounts`, 'Clearing all accounts from keychain')
+  const accounts = await listStoredAccounts()
+  await Promise.all(accounts.map((account) => removeStoredItem(accountStorageKey(account))))
+}
+
+export class KeychainLock {
+  protected locks: Map<
+    string,
+    {
+      // Timestamp in milliseconds when the keychain was last unlocked
+      unlockTime?: number
+      // Number of seconds that the keychain was last unlocked for
+      unlockDuration?: number
+      account: KeychainAccount
+    }
+  > = new Map()
+
+  addAccount(account: KeychainAccount) {
+    this.locks.set(account.address, { account })
+  }
+
+  async unlock(address: string, passphrase: string, duration: number) {
+    Logger.debug(`${TAG}@unlock`, `Unlocking keychain for ${address} for ${duration} seconds`)
+    if (!this.locks.has(address)) {
+      return false
+    }
+    const account = this.locks.get(address)!.account
+    const privateKey = await getStoredPrivateKey(account, passphrase)
+    if (!privateKey) {
+      return false
+    }
+    this.locks.set(address, {
+      account,
+      unlockTime: Date.now(),
+      unlockDuration: duration,
+    })
+    return true
+  }
+
+  isUnlocked(address: string): boolean {
+    if (!this.locks.has(address)) {
+      return false
+    }
+    const { unlockTime, unlockDuration } = this.locks.get(address)!
+    if (unlockDuration === undefined || unlockTime === undefined) {
+      return false
+    }
+
+    if (unlockDuration === 0) {
```

Comment:
Is `unlockDuration === 0` meant to be a special value meaning "unlock indefinitely"? Is this used anywhere/should we allow callers to unlock a signer forever? (And if it is required, should it be documented?)

Reply:
Good point, i added it to the unlock function and this line